SA/design_concept/content/design/design_specification_sum_type_destruction.typ

```diff
@@ -0,0 +1,88 @@
+= Sum Type Destruction
+#let load-drawio-svg(path, ..args) = image.decode(read(path).replace("Text is not SVG - cannot display", ""), ..args)
+
+Sum types consist of a set of constructors, each with a different type. The type
+of a sum type is the union of the types of its constructors.
+
+To work with a value of a sum type, it must be possible to destruct
+the value into its constructor and arguments. Since while developing you
+usually cannot know which constructor was used to create a value, you must
+provide a way to handle all possible constructors. 
+
+A common approach in functional languages is to use _pattern matching_.
+Pattern matching allows to match values against a set of patterns and
+execute a corresponding code. One of the patterns-types usually 
+destructs sum type constructors. 
+
+An example of pattern-matching in Haskell can be seen in
+@pattern-matching-example-haskell, where first a sum type `Expression` is
+defined with two constructors `Value :: Int -> Expression` and
+`Addition :: Expression -> Expression -> Expression`, then a function `calc`
+is defined that pattern-matches a value of type `Expression` against its
+constructors.
+
+#figure(
+    ```hs
+    data Expression where
+      Value :: Int -> Expression
+      Addition :: Expression -> Expression -> Expression
+
+    calc :: Expression -> Int
+    calc e = case e of
+      Value v -> v
+      Addition left right -> calc left + calc right
+    ```,
+    caption: "Example of pattern-matching in Haskell"
+)<pattern-matching-example-haskell>
+
+A more straightforward approach is to generate a function, which, for each
```

Comment:
An easier way to implement this is to generate a function which has a parameter function per constructor of the sum type and an output type. These parameter functions take the type constructor arguments as their input parameters and map them to the output type. This is called a _destruction function_. @destruction-function-example-haskell shows how such a destruction function would look like for the `Expression` example mentioned above.